src/integrator_interface.jl

```diff
@@ -37,34 +37,34 @@ end
 Calculate next step of `integrator`.
 """
 function perform_step!(integrator::DDEIntegrator)
-    # update previous time to extrapolate from current interval
-    integrator.tprev = integrator.t
-
     # update ODE integrator
     integrator.integrator.uprev = integrator.uprev
-    integrator.integrator.tprev = integrator.tprev
-    integrator.integrator.fsalfirst = integrator.fsalfirst
+    integrator.integrator.tprev = integrator.t # to extrapolate from current interval
     integrator.integrator.t = integrator.t
     integrator.integrator.dt = integrator.dt
 
-    # if dt is greater than the minimal lag, then it's explicit so use fixed-point iteration
-    if integrator.dt > minimum(integrator.prob.lags)
+    # separate u to avoid influence of calculations on interpolation and to calculate
+    # residuals
+    if typeof(integrator.u) <: AbstractArray
+        recursivecopy!(integrator.u_cache, integrator.u)
+    else
+        integrator.u_cache = integrator.u
+    end
+    integrator.integrator.u = integrator.u_cache
 
-        # save these values to correct the extrapolation after the last iteration
-        tprev_cache = integrator.tprev
-        t_cache = integrator.t # same as tprev_cache?
-        uprev_cache = integrator.uprev
+    # separate k to avoid influence of calculations on interpolation
+    # constant caches that contain fsalfirst must be updated for interpolation
+    # has to be changed all such caches - move to OrdinaryDiffEq?
+    if typeof(integrator.cache) <: Union{BS3ConstantCache}
```

Comment:
Yes, it will make `k[1] == k[2]` in the first round. However, I'm still not completely sure about which values one would like to use for the interpolation - I mean, I know what we want to achieve but the problem is the correct assignment of `u`, `uprev`, `t`, and `tprev` for `integrator.integrator`. With the current (and the old) setup we have `u == uprev` (by the ODE algorithm) and `t == tprev` (by our assignments) in the first round. In the next rounds `u` is always updated to the newest estimate, `uprev` stays fixed, `t` is fixed to `integrator.t + integrator.dt`, i.e. the time point after this step, and `tprev` is fixed to `integrator.t`, i.e. the current time point.